pkg/clients/sql.go

```diff
@@ -329,189 +210,95 @@ func MySQLVirtualNetworkRuleStatusFromAzure(az mysql.VirtualNetworkRule) azuredb
 //---------------------------------------------------------------------------------------------------------------------
 // PostgreSQLServerClient
 
+// PostgreSQLServerAPI represents the API interface for a MySQL Server client
+type PostgreSQLServerAPI interface {
+	ServerNameTaken(ctx context.Context, s *azuredbv1alpha2.PostgresqlServer) (bool, error)
+	GetServer(ctx context.Context, s *azuredbv1alpha2.PostgresqlServer) (postgresql.Server, error)
+	CreateServer(ctx context.Context, s *azuredbv1alpha2.PostgresqlServer, adminPassword string) error
+	DeleteServer(ctx context.Context, s *azuredbv1alpha2.PostgresqlServer) error
+}
+
 // PostgreSQLServerClient is the concreate implementation of the SQLServerAPI interface for PostgreSQL that calls Azure API.
 type PostgreSQLServerClient struct {
 	postgresql.ServersClient
-	postgresql.FirewallRulesClient
+	postgresql.CheckNameAvailabilityClient
 }
 
 // NewPostgreSQLServerClient creates and initializes a PostgreSQLServerClient instance.
-func NewPostgreSQLServerClient(provider *v1alpha2.Provider, clientset kubernetes.Interface) (*PostgreSQLServerClient, error) {
-	client, err := NewClient(provider, clientset)
-	if err != nil {
-		return nil, fmt.Errorf("failed to create Azure client: %+v", err)
-	}
-
-	postgreSQLServerClient := postgresql.NewServersClient(client.SubscriptionID)
-	postgreSQLServerClient.Authorizer = client.Authorizer
+func NewPostgreSQLServerClient(c *Client) (*PostgreSQLServerClient, error) {
```

Comment:
This applies to a number of the changes to `clients` where they are being refactored to have the `azure.Client` type passed in. I generally prefer passing in something like `[]byte` and then calling to the general `NewClient()` func within `NewPostgreSQLServerClient`. It reduces the code in the managed controllers (slightly, I'll admit), but generally feels cleaner. There is also the option of just doing all of the parsing each time in the actual `New___()` function as is currently being done in the `Redis` client (https://github.com/negz/stack-azure/blob/3b6e68a65f0d1507111937eff1be00a6c899b40b/pkg/clients/redis/redis.go#L45). I may be missing a clear reason why you want to pass in `Client` instead here, so please feel free to push back on my thoughts here :) One reason it may be useful is with something like the AKS cluster where multiple clients are being used. In general, I would prefer consistency with the `New___()` funcs though.

Reply:
I think I agree that it's better to just pass the credential bytes in same as we do elsewhere. Will update.

Reply:
Ergh, actually I just gave this a shot and found it harder to update the AKS controller than I expected due to how its tests work. I still like this direction but would prefer to raise an issue for it rather than try include it in this PR.